src/google/google-calendars/googlecalendarsyncadaptor.cpp

```diff
@@ -53,27 +107,158 @@ void setGCalEventId(KCalCore::Incidence::Ptr event, const QString &id)
+            SOCIALD_LOG_ERROR("unsupported parameter in ex/rdate string:" << exrdatestr);
+            // TODO: support PERIOD formats, or just switch to CalDAV for Google sync...
+        } else if (str.startsWith("TZID=")) {
+            QString tzidstr = str.mid(0, str.indexOf(':'));
+            KTimeZone tz = KSystemTimeZones::zone(tzidstr);
+            str.remove(0, tzidstr.size()+1);
+            QStringList dts = str.split(',');
+            Q_FOREACH (const QString &dtstr, dts) {
+                KDateTime kdt = KDateTime::fromString(dtstr, RFC5545_KDATETIME_FORMAT_NTZC);
+                kdt.setTimeSpec(tz);
+                retn.append(kdt);
+            }
+        } else {
+            SOCIALD_LOG_ERROR("invalid parameter in ex/rdate string:" << exrdatestr);
+        }
+    } else if (str.startsWith(':')) {
+        str.remove(0,1);
+        QStringList dts = str.split(',');
+        Q_FOREACH (const QString &dtstr, dts) {
+            if (dtstr.endsWith('Z')) {
+                // UTC
+                KDateTime kdt = KDateTime::fromString(dtstr, RFC5545_KDATETIME_FORMAT);
+                kdt.setTimeSpec(KDateTime::Spec::UTC());
+                retn.append(kdt);
+            } else {
+                // Floating time
+                KDateTime kdt = KDateTime::fromString(dtstr, RFC5545_KDATETIME_FORMAT_NTZC);
+                kdt.setTimeSpec(KDateTime::Spec::ClockTime());
+                retn.append(kdt);
+            }
+        }
+    } else {
+        SOCIALD_LOG_ERROR("not a valid ex/rdate string:" << exrdatestr);
+    }
+
+    return retn;
+}
+
 QJsonArray recurrenceArray(KCalCore::Event::Ptr event, KCalCore::ICalFormat &icalFormat)
 {
     QJsonArray retn;
+
+    // RRULE
     KCalCore::Recurrence *kcalRecurrence = event->recurrence();
     Q_FOREACH (KCalCore::RecurrenceRule *rrule, kcalRecurrence->rRules()) {
         QString rruleStr = icalFormat.toString(rrule);
         rruleStr.replace("\r\n", "");
         retn.append(QJsonValue(rruleStr));
     }
+
+    // EXRULE
     Q_FOREACH (KCalCore::RecurrenceRule *exrule, kcalRecurrence->exRules()) {
         QString exruleStr = icalFormat.toString(exrule);
         exruleStr.replace("RRULE", "EXRULE");
         exruleStr.replace("\r\n", "");
         retn.append(QJsonValue(exruleStr));
     }
+
+    // RDATE (date)
+    QString rdates;
     Q_FOREACH (const QDate &rdate, kcalRecurrence->rDates()) {
-        retn.append(QJsonValue(QString::fromLatin1("RDATE:%1").arg(rdate.toString("yyyy-MM-dd"))));
+        rdates.append(rdate.toString(RFC5545_QDATE_FORMAT));
+        rdates.append(',');
     }
+    if (rdates.size()) {
+        rdates.chop(1); // trailing comma
+        retn.append(QJsonValue(QString::fromLatin1("RDATE;VALUE=DATE:%1").arg(rdates)));
+    }
+
+    // RDATE (date-time)
+    QString rdatetimes;
+    Q_FOREACH (const KDateTime &rdatetime, kcalRecurrence->rDateTimes()) {
+        rdatetimes.append(rdatetime.toUtc().toString(RFC5545_KDATETIME_FORMAT));
```

Comment:
Is there a better fix?  Eg, can I check: if (rdatetime.spec() == ClockTime) append(rdatetime.toString(No_timezone_format)); instead, perhaps?